actors/miner/src/vesting_state.rs

```diff
@@ -1,54 +1,112 @@
+/// Represents the vesting table state for the miner. It's composed of a `head` (stored inline) and
+/// a tail (referenced by CID).
+#[derive(Serialize_tuple, Deserialize_tuple, Debug, Clone)]
 pub struct VestingFunds {
-    pub funds: Vec<VestingFund>,
+    // The "next" batch of vesting funds:
+    // - If this is None, there are no vesting funds.
+    // - All batches in `tail` are guaranteed to vest after this batch.
+    // - This batch _can_ be empty (have a zero amount) if we've burnt through it (fees &
+    //   penalties).
+    head: Option<VestingFund>,
+    // The rest of the vesting funds, if any.
+    tail: Cid, // Vec<VestingFund>
 }
 
 impl VestingFunds {
-    pub fn new() -> Self {
-        Default::default()
+    pub fn new(store: &impl Blockstore) -> Result<Self, ActorError> {
+        let tail = store
+            .put_cbor(&Vec::<VestingFund>::new(), Code::Blake2b256)
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to construct vesting funds")?;
+        Ok(Self { head: None, tail })
     }
 
-    pub fn unlock_vested_funds(&mut self, current_epoch: ChainEpoch) -> TokenAmount {
-        // TODO: the funds are sorted by epoch, so we could do a binary search here
-        let i = self
-            .funds
-            .iter()
-            .position(|fund| fund.epoch >= current_epoch)
-            .unwrap_or(self.funds.len());
+    pub fn load(
+        &self,
+        store: &impl Blockstore,
+    ) -> Result<impl Iterator<Item = VestingFund>, ActorError> {
+        // NOTE: we allow head to be drawn down to zero through fees/penalties. However, when
+        // inspecting the vesting table, we never want to see a "zero" entry so we skip it in that case.
+        // We don't set it to "none" in that case because "none" means that we have _no_ vesting funds.
+        let head = self.head.as_ref().filter(|h| h.amount.is_positive()).cloned();
+        let tail: Vec<_> = store
+            .get_cbor(&self.tail)
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to load the vesting funds")?
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "missing vesting funds state")?;
+        Ok(itertools::chain(head, tail))
+    }
+
+    fn save(
+        &mut self,
+        store: &impl Blockstore,
+        funds: impl IntoIterator<Item = VestingFund>,
+    ) -> Result<(), ActorError> {
+        let mut funds = funds.into_iter();
+        let head = funds.next();
+        let tail = funds.collect_vec();
+
+        self.tail = store
+            .put_cbor(&tail, Code::Blake2b256)
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to store the vesting funds")?;
+        // We do this second just in case the first operation fails to try to maintain consistent
+        // state.
+        self.head = head;
+        Ok(())
+    }
 
-        self.funds.drain(..i).map(|f| f.amount).sum()
+    pub fn can_vest(&self, current_epoch: ChainEpoch) -> bool {
```

Comment:
can we make this one private? it's slightly weird in that it may return `true` yet have no funds to actually vest. It's more like, "can I run the unlock routine" rather than something you'd care about from the outside